packages/api-contract/src/Abi/index.ts

```diff
@@ -233,7 +262,36 @@ export class Abi {
     });
   };
 
-  #createEvent = (spec: ContractEventSpecLatest, index: number): AbiEvent => {
+  #createEvent = (index: number): AbiEvent => {
+    // TODO TypeScript would narrow this type to the correct version,
+    // but version is `Text` so I need to call `toString()` here,
+    // which breaks the type inference.
+    switch (this.metadata.version.toString()) {
+      case '4':
+        return this.#createEventV4((this.metadata as ContractMetadataV4).spec.events[index], index);
+      default:
+        return this.#createEventV5((this.metadata as ContractMetadataV5).spec.events[index], index);
+    }
+  };
+
+  #createEventV5 = (spec: ContractMetadataV5['spec']['events'][number], index: number): AbiEvent => {
```

Comment:
The type for `spec` is a bit confusing, I think we can narrow it down to `ContractEventSpecV3`

Reply:
instead of `ContractMetadataV5['spec']['events'][number]`

Reply:
I used a local type helper for this to make it less confusing. Otherwise it would be 

```ts
  #createEventV5 = (spec: ContractEventSpecV3): AbiEvent = {}
```

Method stating V5 and expects a V3. Using the helper it's always referring to a `V5`

```ts
  #createEventV5 = (spec: EventOf<ContractMetadataV5>): AbiEvent = {}
```

Internally we have version for some of the contract metadatas fields, like `ContractEventsV1/V2/V3`, `ContractMessageSpecV1/V2/V3` but only the actual top level metadata is properly version. IMHO this was confusing to me. Therefore I like route with the helper here more. Let me know if you disagree. ✌️